dgraph/cmd/dgraphimport/import_client.go

```diff
@@ -0,0 +1,193 @@
+// It opens the BadgerDB at the specified directory and streams all data to the server.
+func streamData(ctx context.Context, dg apiv25.DgraphClient, pdir string, groupId uint32) error {
+	glog.Infof("Opening stream for group %d from directory %s", groupId, pdir)
+
+	// Initialize stream with the server
+	out, err := dg.StreamSnapshot(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to start snapshot stream: %w", err)
+	}
+
+	// Open the BadgerDB instance at the specified directory
+	opt := badger.DefaultOptions(pdir)
+	ps, err := badger.OpenManaged(opt)
+	if err != nil {
+		return fmt.Errorf("failed to open BadgerDB at %s: %w", pdir, err)
+	}
+	defer ps.Close()
+
+	// Send group ID as the first message in the stream
+	glog.V(2).Infof("Sending group ID %d to server", groupId)
+	groupReq := &apiv25.StreamSnapshotRequest{
+		GroupId: groupId,
+	}
+	if err := out.Send(groupReq); err != nil {
+		return fmt.Errorf("failed to send group ID %d: %w", groupId, err)
+	}
+
+	// Configure and start the BadgerDB stream
+	glog.V(2).Infof("Starting BadgerDB stream for group %d", groupId)
+	stream := ps.NewStreamAt(math.MaxUint64)
+	stream.LogPrefix = fmt.Sprintf("Sending P dir (group %d)", groupId)
+	stream.KeyToList = nil
+	stream.Send = func(buf *z.Buffer) error {
+		kvs := &apiv25.KVS{Data: buf.Bytes()}
+		if err := out.Send(&apiv25.StreamSnapshotRequest{
+			Pairs: kvs}); err != nil {
+			return fmt.Errorf("failed to send data chunk: %w", err)
+		}
+		return nil
+	}
+
+	// Execute the stream process
+	if err := stream.Orchestrate(ctx); err != nil {
+		return fmt.Errorf("stream orchestration failed for group %d: %w", groupId, err)
+	}
+
+	// Send the final 'done' signal to mark completion
+	glog.V(2).Infof("Sending completion signal for group %d", groupId)
+	done := &apiv25.KVS{
+		Done: true,
+	}
+
+	if err := out.Send(&apiv25.StreamSnapshotRequest{
+		Pairs: done}); err != nil {
+		return fmt.Errorf("failed to send 'done' signal for group %d: %w", groupId, err)
+	}
+
+	// Wait for acknowledgment from the server
+	ack, err := out.CloseAndRecv()
+	if err != nil {
+		return fmt.Errorf("failed to receive ACK for group %d: %w", groupId, err)
+	}
+	glog.Infof("Group %d: Received ACK with message: %v", groupId, ack.Done)
+
+	return nil
+}
+
+// scanPDirs scans the base path and returns a mapping of group IDs to their
+// corresponding p directory paths. It looks for numbered subdirectories which contain a "p" folder.
+func scanPDirs(basePath string) (map[uint32]string, error) {
+	glog.V(2).Infof("Scanning for p directories in %s", basePath)
+	groupDirs := make(map[uint32]string)
+
+	entries, err := os.ReadDir(basePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read directory %s: %w", basePath, err)
+	}
+
+	for _, entry := range entries {
+		if entry.IsDir() {
+			groupID, err := strconv.ParseUint(entry.Name(), 10, 32)
+			if err == nil {
+				pDir := filepath.Join(basePath, entry.Name(), "p")
+				if _, err := os.Stat(pDir); err == nil {
```

Comment:
do we need to do this here?

Reply:
Yes, we will first check if there are enough p directories, and only then will we start streaming.